### PR TITLE
Add head_mask and decoder_head_mask to FSMT

### LINKED ISSUE #9814: Missing head_mask and decoder_head_mask arguments in encoder-decoder models
# 🚀 Feature request

Following the PRs #9569, #9634 and #9639, there are other encoder-decoder models, which either do not support `head_mask` and `decoder_head_mask` input arguments at all or can be only provided with a single `head_mask` argument used for head masking both in encoder and decoder. It would be, therefore, nice to make this feature uniform over all the decoder-models.

<hr>

**Models:**

| Model | Pytorch | TensorFlow | PR | Copy dependency |
| ------ | :------: | :---------: | :--: | :-----: |
| BERTGeneration | ☑️  | ✖️  | - | - |
| EncoderDecoderModel | ☑️  | ✖️  | - | - |
| FSMT | ☑️  | ✖️  | #9819 | - |
| LED | ☑️ | ☑️ | PT - #9856 ; TF | - |
| ProphetNet | ☑️  | ✖️  | - | - |
| Longformer | ☑️  | ✖️  | #9856  | LED |

## Your contribution

I'm happy to add this feature in the following days, both for PyTorch and TensorFlow models. (Likely in shorter PRs in order not to create large, overwhelming PRs)

<hr>

Reviewers: @patrickvonplaten, @jplu, @sgugger, @LysandreJik, @stas00 .

### PR DESCRIPTION
**Pull request by stancld**
This PR implements `head_mask` and `decoder_head_mask` for FSMT and it is the follow-up to the open issue #9814. 

**Motivation:** This PR is a part of an endeavour to enable the usage of `head_mask` and `decoder_head_mask` for all encoder-decoder transformers following the recent work on BART-like models (#9569).

<hr>

Fixes: https://github.com/huggingface/transformers/issues/9814

Reviewer: @stas00

### CODE REVIEW

**Review by stas00 — APPROVED** [review]
Thank you for syncing fsmt with others, @stancld - including the docs and better markdown!!!

Is there a way to ensure in the future that if things are added to Bart they are added to fsmt too? On modeling level - they are very similar.

**Comment by stancld** [comment]
I know than one can add, for example, a line like this
```
# Copied from transformers.models.bart.modeling_bart.BartAttention with Bart->FSMT
```
before `Attention` module in FSMT. However, this does not copy only additions, but the whole module from BART, which is, in this case, undesirable, I guess, as these modules are a little bit different. But maybe there is another way I am not aware of.

File: tests/test_modeling_fsmt.py [hunk 1]
```diff
@@ -126,7 +134,7 @@ class FSMTModelTest(ModelTesterMixin, GenerationTesterMixin, unittest.TestCase):
     all_generative_model_classes = (FSMTForConditionalGeneration,) if is_torch_available() else ()
     is_encoder_decoder = True
     test_pruning = False
-    test_head_masking = False
+    test_head_masking = True
```
Comment (LysandreJik) [inline tests/test_modeling_fsmt.py:137]: I think this can be removed (as from all other places where it is set to `True`) as this is the default on the superclass.

Imo it makes it more readabe to only have items which are not tested (and which ultimately should be), this way from a quick hover on the test file we can understand what is left to do to have all the common tests coverage. Let me know what you think.

**Review by LysandreJik — APPROVED** [review]
Great, thanks for taking care of it @stancld!

File: src/transformers/models/auto/tokenization_auto.py [hunk 2]
```diff
@@ -170,6 +170,7 @@
     GPT2TokenizerFast = None
     HerbertTokenizerFast = None
     LayoutLMTokenizerFast = None
+    LEDTokenizerFast = None
```
Comment (patrickvonplaten) [inline src/transformers/models/auto/tokenization_auto.py:173]: This was actually recently corrected in another commit this morning - so there might be merge conflicts, but should be easy to resolve

**Review by patrickvonplaten — APPROVED** [review]

File: tests/test_modeling_fsmt.py (hunk 1 above)
Comment (stancld) [reply tests/test_modeling_fsmt.py]: Thanks, that's a good point. I'll create a PR removing all redundant `test_head_masking = True` from other source code files later today.

Comment (stancld) [reply tests/test_modeling_fsmt.py]: Thanks, that's a good point. I'll create a PR removing all redundant test_head_masking = True from other source code files later today.

(Also, I did rebase as `jaxlib` was updated in #9831. Now, it should be, hopefully, fine without any errors :))

**Comment by stas00** [comment]
@LysandreJik, @patrickvonplaten - how can we make sure fsmt gets tracked and synced with all the bart-family changes? while the tokenizer is different, the model is ~95% identical.

**Comment by LysandreJik** [comment]
as @stancld said, we can do that with some statements of the following kind:

```
# Copied from transformers.models.bart.modeling_bart.BartAttention with Bart->FSMT
```

The difference between the BART and FSMT implementation of the targeted object must only be the "BART" occurrences that change to "FSMT". @sgugger can tell you more about it.

**Comment by stas00** [comment]
Thank you, @LysandreJik

I think this is really a question to @patrickvonplaten  - who I remember was planning to refactor FSMT to match what he did for Bart. So if this is still planned, Patrick perhaps you could add this item to the agenda - keeping FSMT in sync with the Bart-family (modeling only - tokenizer is similar to xlm).

It might help to treat FSMT as Bart with the main difference of it having a dual vocab and no tied weights - and a few layers that are different - but identical otherwise. (again for the model only).

**Comment by patrickvonplaten** [comment]
> I think this is really a question to @patrickvonplaten - who I remember was planning to refactor FSMT to match what he did for Bart. So if this is still planned, Patrick perhaps you could add this item to the agenda - keeping FSMT in sync with the Bart-family (modeling only - tokenizer is similar to xlm).

Yes, the FSTM / ProphetNet refactor is still on my ToDo List (think next week is reasonable). After the refactor I'll try to add as many # Copied from statements to keep the models in sync. Nevertheless, this PR can be merged as it is now!

Great work @stancld
